### PR TITLE
saveclip at backend error handling

### LINKED ISSUE #1459: Let clip saving errors bubble up
Didn't happen on production in the last time, but we noticed that those errors are caught too early.

### PR DESCRIPTION
**Pull request by mritunjaygoutam12**
Fixes #1459

### CODE REVIEW

**Comment by mritunjaygoutam12** [comment]
@Gregoor review it, please

**Comment by Gregoor** [comment]
I will review it tomorrow 🤞

File: web/src/services/api.ts [hunk 1]
```diff
@@ -67,7 +67,7 @@ export default class API {
       return;
     }
     if (response.status >= 400) {
-      throw new Error(await response.text());
+      throw response;
```
Comment (Gregoor) [inline web/src/services/api.ts]: why not wrap it in an error object?

File: server/src/lib/clip.ts [hunk 2]
```diff
@@ -120,64 +120,71 @@ export default class Clip {
...
+        path: clipFileName,
+        sentence,
+        sentenceId: headers.sentence_id,
+      });
+      await Awards.checkProgress(client_id);
 
-    response.json(filePrefix);
+      response.json(filePrefix);
+    } catch (error) {
+      console.log(error);
```
Comment (Gregoor) [inline server/src/lib/clip.ts]: let's use `console.error` instead

File: server/src/lib/clip.ts [hunk 3]
```diff
@@ -120,64 +120,71 @@ export default class Clip {
...
+      response.status(error.statusCode || 500).json(error);
```
Comment (Gregoor) [inline server/src/lib/clip.ts]: let's just return a generic error message here, error will be meaningless to the client anyway.

Comment (mritunjaygoutam12) [reply server/src/lib/clip.ts]: Ok, then just send some error message then in below speak.tsx only check for that error message to differentiate between the errors ??

File: server/src/lib/clip.ts (hunk 2 above)
Comment (mritunjaygoutam12) [reply server/src/lib/clip.ts]: :+1:

File: web/src/services/api.ts (hunk 1 above)
Comment (mritunjaygoutam12) [reply web/src/services/api.ts]: :+1:

File: server/src/lib/clip.ts (hunk 3 above)
Comment (mritunjaygoutam12) [reply server/src/lib/clip.ts]: let me do this

Comment (Gregoor) [reply server/src/lib/clip.ts]: I think it'd be best if we just console.error'd it here and output something generic, as the client doesn't need to know about the internals here.

Comment (mritunjaygoutam12) [reply server/src/lib/clip.ts]: :+1:

**Comment by mritunjaygoutam12** [comment]
conflict resolved :relaxed:

File: web/src/services/api.ts [hunk 4]
```diff
@@ -66,8 +66,8 @@ export default class API {
       location.reload();
       return;
     }
-    if (response.status >= 400) {
-      throw new Error(await response.text());
+    if (response.statusText === 'Error in saveClip server') {
```
Comment (Gregoor) [inline web/src/services/api.ts]: I don't think we should replace the generic error handling here. Also the generic case would also cover this case wouldn't it?

File: server/src/lib/clip.ts [hunk 5]
```diff
@@ -123,66 +123,75 @@ export default class Clip {
...
+      });
+      await Awards.checkProgress(client_id);
 
     response.json(filePrefix);
 
     Basket.sync(client_id).catch(e => console.error(e));
+      response.json(filePrefix);
+    } catch (error) {
+      console.error(error);
+      response.statusMessage = 'Error in saveClip server';
```
Comment (Gregoor) [inline server/src/lib/clip.ts]: maybe name it something like `save_clip_error` instead, and then check directly for it in `speak.tsx`?

**Comment by mritunjaygoutam12** [comment]
little new approach. Steps:
>>>  Catching error in backend clip.ts and responded it with a status code and message. status code because it can be handled by previous generic error handling in services/api.ts

>>>Throwing separate new error for save_clip_error  in api.ts because of previous error throw hole error as text so I am unable to operate on the error in speak.tsx

**Comment by Gregoor** [comment]
Great, looks good to me. Thanks for your hard work and patience with me :)

**Pull request merged**